beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -85,60 +86,41 @@ impl<E: EthSpec> RangeBlockComponentsRequest<E> {
+            let expected_column_indices = expected_column_indices.clone();
+            self.into_responses_with_custody_columns(spec, expected_column_indices)
         } else {
-            self.into_responses_with_blobs()
+            Ok(self.into_responses_with_blobs())
         }
     }
 
-    fn into_responses_with_blobs(self) -> Result<Vec<RpcBlock<E>>, String> {
+    fn into_responses_with_blobs(self) -> Vec<RpcBlock<E>> {
         let RangeBlockComponentsRequest { blocks, blobs, .. } = self;
 
-        // There can't be more more blobs than blocks. i.e. sending any blob (empty
-        // included) for a skipped slot is not permitted.
-        let mut responses = Vec::with_capacity(blocks.len());
-        let mut blob_iter = blobs.into_iter().peekable();
-        for block in blocks.into_iter() {
-            let mut blob_list = Vec::with_capacity(E::max_blobs_per_block());
-            while {
-                let pair_next_blob = blob_iter
-                    .peek()
-                    .map(|sidecar| sidecar.slot() == block.slot())
-                    .unwrap_or(false);
-                pair_next_blob
-            } {
-                blob_list.push(blob_iter.next().ok_or("Missing next blob".to_string())?);
-            }
+        let mut blobs_by_block = HashMap::<Hash256, Vec<Arc<BlobSidecar<E>>>>::new();
 
-            let mut blobs_buffer = vec![None; E::max_blobs_per_block()];
-            for blob in blob_list {
-                let blob_index = blob.index as usize;
-                let Some(blob_opt) = blobs_buffer.get_mut(blob_index) else {
-                    return Err("Invalid blob index".to_string());
-                };
-                if blob_opt.is_some() {
-                    return Err("Repeat blob index".to_string());
-                } else {
-                    *blob_opt = Some(blob);
-                }
-            }
-            let blobs = VariableList::from(blobs_buffer.into_iter().flatten().collect::<Vec<_>>());
-            responses.push(RpcBlock::new(None, block, Some(blobs)).map_err(|e| format!("{e:?}"))?)
+        for blob in blobs {
+            let block_root = blob.block_root();
+            blobs_by_block.entry(block_root).or_default().push(blob);
         }
 
-        // if accumulated sidecars is not empty, throw an error.
-        if blob_iter.next().is_some() {
-            return Err("Received sidecars that don't pair well".to_string());
+        // Now iterate all blocks ensuring that the block roots of each block and data column match,
+        // plus we have columns for our custody requirements
+        let mut rpc_blocks = Vec::with_capacity(blocks.len());
+
+        for block in blocks {
+            let block_root = get_block_root(&block);
+            let blobs = blobs_by_block.remove(&block_root).map(VariableList::from);
+            rpc_blocks.push(RpcBlock::new_unchecked(block_root, block, blobs, None));
```

Comment:
I think using the unchecked variant here might result in an inconsistent state.

Consider the example you used, the peer/peers have sent us a valid block for a block_root `x` which contains 1 `kzg_commitment`. The peer sends us 2 blobs for this block root which do not match with the `kzg_commitment` in the block. The malicious blobs pass the kzg verification in isolation, they just doesn't match with the block commitment.

Now the unchecked `RpcBlock` is sent for processing.  The block goes through the validation pipeline
I couldn't find anywhere we check that the block and blob commitments match up. Correct me if i'm wrong here though.

We should potentially do this check when we do`RpcBlock::deconstruct`

Reply:
> I couldn't find anywhere we check that the block and blob commitments match up

In sync when doing a blobs request we check the inclusion proof, here

https://github.com/sigp/lighthouse/blob/40d3423193e6f97f44c42eb3673af607627c8f7f/beacon_node/network/src/sync/network_context/requests/blobs_by_root.rs#L53